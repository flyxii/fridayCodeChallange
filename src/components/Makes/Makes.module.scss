@use 'libs/styles/dataWrapper.scss' with (
  $width: 25%,
);
@import "libs/styles/base.scss";

.make{
  @extend %bodyText;
  color: 333;
  padding: 16px 16px;
  cursor: pointer;

  &:hover{
    background-color: $lightGrey;
  }

  &.selected{
    background-color: $purple;
    color: #fff;
  }
}