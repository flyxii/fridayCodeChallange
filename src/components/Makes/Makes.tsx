/**
 * Component to fetch the list of makes and display the return data to end user
 */

import React, { useContext } from "react";
import useSWR from "swr";
import { AppContext } from "store/context";
import { Action } from "store/type";
import Loading from "libs/images/Loading.svg";
import styles from "./Makes.module.scss";

const Makes: React.FC = () => {
  /**
   * Connect to the store
   */
  const { state, dispatch } = useContext(AppContext);

  /**
   * Fetch makes data from api
   */
  const { data, error } = useSWR<string[], Error>("http://localhost:8080/api/makes");

  /**
   * Show loading icon if waiting for data to come or error return from api
   */
  if (error || !data)
    return (
      <div className={styles.error}>
        <img src={Loading} className={styles.loading} alt="Loading" data-testid="loading" />
      </div>
    );

  /**
   * Show the list of makes
   */
  return (
    <div className={styles.wrapper}>
      <div className={styles.title}>Makes</div>
      <div className={styles.scrollWrapper}>
        {data.map((item, i) => (
          <div
            className={state.make === item ? styles.make + " " + styles.selected : styles.make}
            key={i}
            onClick={() => {
              dispatch({ type: Action.UPDATE_MAKE, payload: item });
              dispatch({ type: Action.UPDATE_MODEL, payload: "" });
            }}>
            {item}
          </div>
        ))}
      </div>
    </div>
  );
};

export default Makes;
