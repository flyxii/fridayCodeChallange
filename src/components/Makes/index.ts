import Makes from "./Makes";

export default Makes;
