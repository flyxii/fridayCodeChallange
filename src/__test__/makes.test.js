import * as React from "react";
import { SWRConfig } from "swr";
import fetch from "jest-fetch-mock";
import { render, waitFor } from "@testing-library/react";
import Makes from "components/Makes";
import fetcher from "libs/fetcher";

const App = () => (
  <SWRConfig value={{ dedupingInterval: 0, shouldRetryOnError:false, fetcher: fetcher }}>
    <Makes />
  </SWRConfig>
);

describe("<Makes />", () => {
  beforeEach(() => {
    fetch.resetMocks();
    fetch.doMock();
    jest.useFakeTimers();
  });
  afterEach(() => {
    jest.useRealTimers();
  });

  it("renders without crashing", async () => {
    const { container } = render(<App />);
    await waitFor(() => {
      expect(container.firstChild).toBeInTheDocument();
    });
  });

  it("shows laoding icon if fetch reject", async () => {
    fetch.mockReject(() => Promise.reject("server is down"));
    const { getByTestId, asFragment } = render(<App />);

    await waitFor(() => {
      const loading = getByTestId(/loading/i);
      expect(loading).toBeInTheDocument();
      expect(fetch).toHaveBeenCalledTimes(1);
      expect(fetch).toHaveBeenCalledWith("http://localhost:8080/api/makes", undefined);
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it("shows loading icon in loading stage", async () => {
    fetch.mockResponseOnce("");
    const { getByTestId, asFragment } = render(<App />);

    await waitFor(() => {
      const loading = getByTestId(/loading/i);
      expect(loading).toBeInTheDocument();
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it("shows data if fetch success", async () => {
    const mockResponse = ["BMW", "Ford"];

    fetch.mockResponseOnce(JSON.stringify(mockResponse));
    const { findAllByText, asFragment } = render(<App />);

    const bmw = await findAllByText("BMW");
    const ford = await findAllByText("Ford");

    expect(bmw).toHaveLength(1);
    expect(ford).toHaveLength(1);
    expect(asFragment()).toMatchSnapshot();
  });
});
