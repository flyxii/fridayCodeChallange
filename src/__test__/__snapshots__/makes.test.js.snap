// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Makes /> shows data if fetch success 1`] = `
<DocumentFragment>
  <div
    class="wrapper"
  >
    <div
      class="title"
    >
      Makes
    </div>
    <div
      class="scrollWrapper"
    >
      <div
        class="make"
      >
        BMW
      </div>
      <div
        class="make"
      >
        Ford
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Makes /> shows laoding icon if fetch reject 1`] = `
<DocumentFragment>
  <div
    class="error"
  >
    <img
      alt="Loading"
      class="loading"
      data-testid="loading"
      src="Loading.svg"
    />
  </div>
</DocumentFragment>
`;

exports[`<Makes /> shows loading icon in loading stage 1`] = `
<DocumentFragment>
  <div
    class="error"
  >
    <img
      alt="Loading"
      class="loading"
      data-testid="loading"
      src="Loading.svg"
    />
  </div>
</DocumentFragment>
`;
